src/lib/Imp.hs

```diff
@@ -1074,6 +1053,40 @@ _deviceFromCallingConvention cc = case cc of
   MCThreadLaunch    -> CPU
   CUDAKernelLaunch  -> GPU
 
+-- === Determining buffer sizes and offsets using polynomials ===
+
+type IndexStructure = EmptyAbs IdxNest :: E
+
+computeElemCount :: (Emits n, Builder m) => IndexStructure n -> m n (Atom n)
+computeElemCount idxNest = do
+  polySize <- liftImmut $ elemCountCPoly idxNest
+  A.emitCPoly polySize
+
+computeOffset :: (Emits n, Builder m) => IndexStructure n -> [AtomName n] -> m n (Atom n)
+computeOffset (Abs Empty UnitE) [] = return $ IdxRepVal 0
+computeOffset (Abs (Nest b bs) UnitE) (i:is) = do
+  rhsElemCounts <- liftImmut $ refreshBinders b \(b':>_) s -> do
+    rest' <- applySubst s $ EmptyAbs bs
+    Abs b' <$> elemCountCPoly rest'
+  significantOffset <- A.emitCPoly $ A.sumC i rhsElemCounts
+  remainingIdxStructure <- applySubst (b@>i) (EmptyAbs bs)
+  otherOffsets <- computeOffset remainingIdxStructure is
+  iadd significantOffset otherOffsets
+computeOffset _ _ = error "zip error"
+
+elemCountCPoly :: (EnvExtender m, EnvReader m, Immut n)
+               => IndexStructure n -> m n (A.ClampPolynomial n)
+elemCountCPoly (Abs bs UnitE) = case bs of
+  Empty -> return $ A.liftPoly $ A.emptyMonomial
+  Nest b rest -> do
+    size <- A.indexSetSizeCPoly $ binderType b
+    rhsElemCounts <- refreshBinders b \(b':>_) s -> do
+      rest' <- applySubst s $ Abs rest UnitE
+      Abs b' <$> elemCountCPoly rest'
+    withFreshBinder NoHint IdxRepTy \b' -> do
+      let sumPoly = A.sumC (binderName b') (sink rhsElemCounts)
+      return $ A.psubst (Abs b' sumPoly) size
```

Comment:
It used to be the case that we took advantage of the mutual recursion between counting elements and offsets (in that the total number of elements is obtained by substituting the upper bound for the offset). Why can't we do that anymore?

Reply:
We could still do that. I wrote it this way because it was a bit clearer to me, even though there's some slight duplication. The other way was neat, but when I first read it, it took me a while to realize the trick. I don't have strong feelings either way.